### DISCUSSION TITLE
How to set the initial z-value?

### DISCUSSION THREAD

**Discussion by whY-note** [bitcraze/discussions#1628] [Q&A] [answered]
Hello, when my crazyflie begin to fly, its initial z-value is about -18 meters.
It is very strange. 
Could you tell me why the initial z-value is about -18 meters? 
And could you teach me how to set the initial z-value?

Thank you!
Your prompt reply will be highly appreciated.

**Comment by gemenerik** [comment]
I need a bit more information about the expansion decks you might be using with your Crazyflie. However, assuming you’re not using any positioning decks, the Crazyflie defaults to the complementary filter for estimation, which relies primarily on the barometer for altitude initialization. If you’re seeing an initial z-value of -18 meters, and you're not actually 18 meters below sea level or in a depressurized environment, it’s likely due to barometer bias.

For instance, the Crazyflie on my desk currently reads 130 meters below sea level, even though we're at about 7 meters above sea level. This bias doesn’t affect relative accuracy, though: if I move the Crazyflie up by 1 meter, it reads 129 meters below sea level, indicating that while absolute values may be off, relative changes remain accurate. To address this in your setup, you could add a correction offset to the barometer readings in the `position_estimator_altitude.c` file within the Crazyflie firmware to compensate for the bias. This adjustment should help achieve a more realistic initial altitude.

**Reply by whY-note** [reply]
Thanks a lot!

My Crazyflie is not equipped with any expansion decks, as you assumed.
After I add a correction offset to the barometer readings in the position_estimator_altitude.c file, the initial z-value became nearly zero.
But when I turned off the Crazyflie and turned it on again, the initial z-value became nearly -3 at the same place.
It seems that the initial z-values are different every time I turn on the Crazyflie. It is very strange.

Fortunately, its relative accuracy is accurate.
However, I need to  set a specific flight trajectory. So I want the initial z-value is accurate. 
For example, if I want my Crazyflie fly to the point (0,0,1) and the initial z-value is -20, then it will fly 21 metres upwards. On the contrary, if I want my Crazyflie fly to the point (0,0,1) and the  initial z-value is 0, then it will fly just 1 metres upwards.
That's why I need an accurate initial z-value.

Could you help me with this problem?
Your prompt reply will be highly appreciated.

**Reply by gemenerik** [reply, marked as answer]
To achieve a more consistent initial z-value, you can try dynamically setting the barometer bias upon initialization. During startup, read the current barometer value and use it as a baseline or bias for altitude calculations. This way, each time the Crazyflie initializes, it will account for the ambient pressure and start with a z-value close to zero, minimizing discrepancies across power cycles.

Keep in mind that without any expansion decks, the Crazyflie will likely experience noticeable drift in both the X and Y axes due to the lack of external positioning references. Additionally, the barometer alone has limited precision on the Z-axis. For more accurate position tracking, you might want to consider adding positioning decks in the future.

**Reply by whY-note** [reply]
Thank you very much! Your answer is very helpful.

I will buy a deck soon.
